### ISSUE TITLE
Please, give us "signs of success"

### ISSUE THREAD

**Issue by BugsDigger** [zeffy/wufuc#168]
Hi, many thanks for your soft !

> wufuc version (can be viewed in Add & Remove Programs).
Installed from zip 1.0.1.201-a52637d4, no info in "Add & Remove Programs".
> Operating system version and Service Pack
Win7/64 with all (available) updates.

After installation I don't see any "signs of success" (except of absence of annoying message in question at moment :-)).

I see a difference (~1 hour) in timestamps of wufuc64.dll if I install from msi or from zip. Is this important?

Depending on istallation source (msi or zip) I see in log different messages like "can not create mutex" (with msi) or nothing (except of msg about stopping of the service during bat-installation). I see sheduled task wufuc (using CCleaner utility) and rundll32 with wufuc64.dll in memory. Does it mean that installation is OK?

Thank you.

**Comment by zeffy** [comment]
>After installation I don't see any "signs of success" (except of absence of annoying message in question at moment :-)).

If you check the log file it will say if patching was successful or not, and if you can open Windows Update and click "Check for updates" in the side pane without getting the unsupported hardware popup, then it is working correctly.

>I see a difference (~1 hour) in timestamps of wufuc64.dll if I install from msi or from zip. Is this important?

On the file itself? No.

>Depending on istallation source (msi or zip) I see in log different messages like "can not create mutex" (with msi) or nothing (except of msg about stopping of the service during bat-installation). I see sheduled task wufuc (using CCleaner utility) and rundll32 with wufuc64.dll in memory. Does it mean that installation is OK?

Usually that means wufuc is already running.

**Comment by BugsDigger** [comment]
Thank you for feedback.
Unfortunately I've got this damned MS message regardless of the fact that I see your patch in memory... :-(
Log is empty (no any info relative to "patching was successful or not").

Does your patch depend on system dll(s) version(s)? (And if yes, with what of them it was tested?)

**Comment by zeffy** [comment]
>Does your patch depend on system dll(s) version(s)? (And if yes, with what of them it was tested?)

No, it's mostly self contained other than using some standard windows APIs that have been around since Vista or earlier.

Don't know what could be causing that, I've seen a couple people report similar issues. You could try killing the existing rundll32.exe process and run the wufuc task manually from the Task Scheduler.

**Comment by BugsDigger** [comment]
Tried this from cmd with admin rights:
schtasks /Run /TN "wufuc.{72eee38b-9997-42bd-85d3-2dd96da17307}"
Success

dll appears in memory, log is empty. Check for updates does not show its nasty message right now (probably because there are no updates at moment?.. Or because patch is really working now?..)

In what cases wufuc task log something? On every start/err/success? (Why my log is empty?) Is there any special cmd line key for verbose logging?

**Comment by zeffy** [comment]
by default it logs errors and upon patch success. honestly I don't know what would cause it to not log anything, other than perhaps permission issues. 

one thing you can try is get [DebugView](https://docs.microsoft.com/en-us/sysinternals/downloads/debugview) and see if anything gets output there. wufuc is supposed to fall back to OutputDebugString in the event that writing to the log file fails.

**Comment by zeffy** [comment]
Yeah, that's really weird. I haven't encountered anything like this during testing. 

I wish I could look into it more, but I've been really tied up recently and haven't had a lot of free time. Don't really know when I'll have a chance to, unfortunately.

**Comment by BugsDigger** [comment]
Hi zeffy.
I see that this statement is false in my case:

  ```
 if ( (s_bWindowsSevenSP1 &&
        vercmp(lpffi->dwProductVersionMS, lpffi->dwProductVersionLS, 7, 6, 7601, 23714) != -1) ...
```

I have 23453 instead of 23714 (and there is no log message about  it, BTW).
I've specially updated it, so I believe this is the latest version. 

Strange, there are 2 dlls: wuaueng.dll (v.7.6.7601.24085) and wuaueng2.dll (v.7.6.7601.23453). Debugging shows me internal name "wuaueng.dll" and v.23453 ...

I see the only place with pattren "FFF3 4883EC?? 33DB 391D???????? 7508 8B05????????" in wuaueng.dll (and no such pattren in wuaueng2.dll). Do you think that it is safe to patch this version?

About empty log: is this a buffering issue? (FlushFileBuffers in UTF8WriteFile can help)

**Comment by zeffy** [comment]
@BugsDigger wuaueng2.dll comes from UpdatePack7R2, wufuc is not compatible with it by default, you have to merge the included Restore_wuaueserv.reg before it will work correctly. This is because it has its own patched .dll that it installs and changes the registry to use it instead. Once you fix the registry entry, it will point (correctly) to wuaueng.dll (v.7.6.7601.24085) and that if statement will evaluate to true.

In regarded to FlushFileBuffers, I believe the file buffers should be flushed regardless when the file handle is closed at process exit, but you may be right. I will have to look into it.

**Comment by BugsDigger** [comment]
@zeffy Aha, now all is working as expected, many thanks! :-) 
I see no mention of Restore_wuaueserv.reg in docs. Could you please point me out where it is mentioned?

About flushing log file: process enters into waiting of events, so it has no chance to save contents on disk for too long time. From log viewer's (user's) point of view it looks as empty log.

**Comment by zeffy** [comment]
>Aha, now all is working as expected, many thanks! :-)

Nice!

>Could you please point me out where it is mentioned?

https://github.com/zeffy/wufuc/blob/micro/FAQ.md#how-to-restore-windows-update-to-its-default-configuration :) but to be fair the FAQ file kind of hidden away.

>About flushing log file: process enters into waiting of events, so it has no chance to save contents on disk for too long time. From log viewer's (user's) point of view it looks as empty log.

You have a point, it will still get flushed when you restart but like you say, that isn't much use to a user who is trying to figure out if it is working or not! I will try to get this added when I have a chance.

**Comment by BugsDigger** [comment]
@zeffy > FAQ
Ah, yes, I see now.
(The only problem that I had not other patches ... :-))
